src/main/scala/is/hail/variant/MatrixTable.scala

```diff
@@ -914,9 +914,11 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) {
   def annotateRowsVDS(right: MatrixTable, root: String): MatrixTable =
     orderedRVDLeftJoinDistinctAndInsert(right.value.rowsRVD(), root, product = false)
 
-  def count(): (Long, Long) = (countRows(), numCols)
+  def count(): (Long, Long) = (countRows(), countCols())
 
-  def countRows(): Long = partitionCounts().sum
+  def countRows(): Long = Interpret(TableCount(MatrixRowsTable(ast)))
+
+  def countCols(): Long = Interpret(TableCount(MatrixColsTable(ast)))
```

Comment:
This is going to turn a ~free operation into one that parallelizes the cols table and then counts, right? can we make MatrixColsTable store partitionCount information to prevent that from happening?

Reply:
it's definitely not free! This requires evaluating the MatrixValue, which means a huge number of optimizations won't get run. This should get pushed down all the way to a MatrixRead if possible.